InnerEye/ML/model_training.py

```diff
@@ -3,186 +3,241 @@
+    # training patches is not stable enough. Going by the validation loss somehow works for the Prostate model, but
+    # not for the HeadAndNeck model.
+    best_checkpoint_callback = ModelCheckpoint(dirpath=str(config.checkpoint_folder),
+                                               # filename=BEST_CHECKPOINT_FILE_NAME,
+                                               # monitor=f"{VALIDATION_PREFIX}{MetricType.LOSS.value}",
+                                               # save_top_k=1,
+                                               save_last=True)
+    # Recovery checkpoints: {epoch} will turn into a string like "epoch=1"
+    # Store 1 recovery checkpoint every recovery_checkpoint_save_interval epochs. Due to a bug in Lightning, this
+    # will still write alternate files recovery.ckpt and recovery-v0.ckpt, which are cleaned up later in
+    # cleanup_checkpoint_folder
+    recovery_checkpoint_callback = ModelCheckpoint(dirpath=str(config.checkpoint_folder),
+                                                   filename=RECOVERY_CHECKPOINT_FILE_NAME,
+                                                   period=config.recovery_checkpoint_save_interval
+                                                   )
+
+    num_gpus = torch.cuda.device_count() if config.use_gpu else 0
+    logging.info(f"Number of available GPUs: {num_gpus}")
+    if config.max_num_gpus >= 0 and config.max_num_gpus < num_gpus:
+        num_gpus = config.max_num_gpus
+        logging.info(f"Restricting the number of GPUs to {num_gpus}")
+    # Accelerator should be "ddp" when running large models in AzureML (when using DDP_spawn, we get out of GPU memory).
+    # For unit tests, only "ddp_spawn" works
+    accelerator = "ddp" if num_gpus > 1 else None
+    logging.info(f"Using {num_gpus} GPUs with accelerator '{accelerator}'")
+    storing_logger = StoringLogger()
+    tensorboard_logger = TensorBoardLogger(save_dir=str(config.logs_folder), name="Lightning", version="")
+    loggers = [storing_logger, tensorboard_logger, AzureMLLogger()]
+    # This leads to problems with run termination.
+    # if not is_offline_run_context(RUN_CONTEXT):
+    #     mlflow_logger = MLFlowLogger(experiment_name=RUN_CONTEXT.experiment.name,
+    #                                  tracking_uri=RUN_CONTEXT.experiment.workspace.get_mlflow_tracking_uri())
+    #     # The MLFlow logger needs to get its ID from the AzureML run context, otherwise there will be two sets of
+    #     # results for each run, one from native AzureML and one from the MLFlow logger.
+    #     mlflow_logger._run_id = RUN_CONTEXT.id
+    #     loggers.append(mlflow_logger)
+
+    trainer = Trainer(default_root_dir=str(config.outputs_folder),
+                      deterministic=True,
+                      accelerator=accelerator,
+                      max_epochs=config.num_epochs,
+                      num_sanity_val_steps=0,  # Otherwise a small number of validation steps is run before first train
```

Comment:
shouldn't we set this to a non-zero number? If there is an error in the validation path, it will trigger early on instead of waiting for full training epoch to end. In my experience we usually gain more time than we lose in one small val step.

Reply:
made it a parameter.